sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/DefaultSynchronousMetricStorage.java

```diff
@@ -208,25 +232,79 @@ public MetricData collect(
       aggregatorHandles = this.aggregatorHolder.aggregatorHandles;
     }
 
+    List<T> points;
+    if (memoryMode == REUSABLE_DATA) {
+      reusableResultList.clear();
+      points = reusableResultList;
+    } else {
+      points = new ArrayList<>(aggregatorHandles.size());
+    }
+
+    // In DELTA aggregation temporality each Attributes is reset to 0
+    // every time we perform a collection (by definition of DELTA).
+    // In IMMUTABLE_DATA MemoryMode, this is accomplished by removing all aggregator handles
+    // (into which the values are recorded) effectively starting from 0
+    // for each recorded Attributes.
+    // In REUSABLE_DATA MemoryMode, we strive for zero allocations. Since even removing
+    // a key-value from a map and putting it again on next recording will cost an allocation,
+    // we are keeping the aggregator handles in their map, and only reset their value once
+    // we finish collecting the aggregated value from each one.
+    // The SDK must adhere to keeping no more than maxCardinality unique Attributes in memory,
+    // hence during collect(), when the map is at full capacity, we try to clear away unused
+    // aggregator handles, so on next recording cycle using this map, there will be room for newly
+    // recorded Attributes. This comes at the expanse of memory allocations. This can be avoided
+    // if the user chooses to increase the maxCardinality.
+    if (memoryMode == REUSABLE_DATA && reset) {
+      if (aggregatorHandles.size() >= maxCardinality) {
+        aggregatorHandles.forEach(
+            (attribute, handle) -> {
+              if (!handle.isValuesRecorded()) {
+                aggregatorHandles.remove(attribute);
```

Comment:
We should return these to the pool.
```suggestion
                AggregatorHandle<T, U> removedHandle = aggregatorHandles.remove(attribute);
                aggregatorHandlePool.offer(removedHandle);
```

Reply:
Not is REUSABLE_DATA mode, since when ever you add (`offer`) or remove (`pop`) from the `ConcurrentLinkedQueue` (pool) you allocate a `Node`, which makes it `O(#Attributes)`.
The only reason to use the pool here is if creating 2 `Node`s is cheaper than allocating a handler. 
We need to keep in mind that in REUSABLE_DATA we're not using the pool at all, since we accumulating Attribute sets. 
WDYT

Reply:
My intuition tells me that creating 1 handler is more expensive than 2 nodes in the pool. Still, its a nice simplifying assumption to asset that REUSABLE_DATA doesn't use the pool (its ironic that reusable mode does not reuse handlers). I think we'll ultimately want to refactor this to more cleanly separate the behavioral differences between the implementations. For now, let's _not_ return it to the pool so that a future refactor isn't tempted to bring the object pool with the delta reusable implementation.

Reply:
I have safeguards against future refactor in part 2 or 3 of this splitted PR - a test making sure allocation rate is 99.98% vs immutable, both for DELTA and CUMULATIVE.